### ISSUE TITLE
Multiplicity models do not appear in drop-down in Custom Model > Sum|Multi (Trac #473)

### ISSUE THREAD

**Issue by smk78** [SasView/sasview#607]
Reported by Saeed Mohan (saeed.mohan@ipleiria.pt):

I was attempting to add in my own custom model to Sasview, which is based on the !UnifiedPowerRg model. I thought the easiest thing to do would be to use the Sum Multi|(p1, p2) function under Fitting > Edit Custom Model, however I am not able to find this model in the options list.

Checked by SMK. He is correct.

What is more, the same bug exists in v3.0.0 and v2.2.1!!!

PDB checked and it appears that no multiplicity models appear in the dropdown (eg. RPA10, onion, etc).  This is probably working as designed since I suspect the interface was not obvious for a multiplicity model?  Interestingly there is also a very strange behavior that the shape independent -> Absolute_Power_Law seems to have been renamed in this dropdown to !PowerLawAbsModel ... perhaps due to the space in the name? 

Migrated from http://trac.sasview.org/ticket/473
```json
{
    "status": "new",
    "changetime": "2019-03-02T00:16:28",
    "_ts": "2019-03-02 00:16:28.434007+00:00",
    "description": "Reported by Saeed Mohan (saeed.mohan@ipleiria.pt):\n\nI was attempting to add in my own custom model to Sasview, which is based on the !UnifiedPowerRg model. I thought the easiest thing to do would be to use the Sum Multi|(p1, p2) function under Fitting > Edit Custom Model, however I am not able to find this model in the options list.\n\nChecked by SMK. He is correct.\n\nWhat is more, the same bug exists in v3.0.0 and v2.2.1!!!\n\nPDB checked and it appears that no multiplicity models appear in the dropdown (eg. RPA10, onion, etc).  This is probably working as designed since I suspect the interface was not obvious for a multiplicity model?  Interestingly there is also a very strange behavior that the shape independent -> Absolute_Power_Law seems to have been renamed in this dropdown to !PowerLawAbsModel ... perhaps due to the space in the name? ",
    "reporter": "smk78",
    "cc": "",
    "resolution": "",
    "workpackage": "SasModels Infrastructure",
    "time": "2015-11-10T12:00:02",
    "component": "SasView",
    "summary": "Multiplicity models do not appear in drop-down in Custom Model > Sum|Multi",
    "priority": "major",
    "keywords": "",
    "milestone": "SasView 5.1.0",
    "owner": "",
    "type": "defect"
}
```

**Comment by butlerpd** [comment]
Trac update at `2015/11/28 21:29:47`: 
* **butler** changed description from:

> Reported by Saeed Mohan (saeed.mohan@ipleiria.pt):
> 
> I was attempting to add in my own custom model to Sasview, which is based on the UnifiedPowerRg model. I thought the easiest thing to do would be to use the Sum Multii|(p1, p2) function under Fitting > Edit Custom Model, however I am not able to find this model in the options list.
> 
> Checked by SMK. He is correct.
> 
> What is more, the same bug exists in v3.0.0 and v2.2.1!!!

to:

> Reported by Saeed Mohan (saeed.mohan@ipleiria.pt):
> 
> I was attempting to add in my own custom model to Sasview, which is based on the !UnifiedPowerRg model. I thought the easiest thing to do would be to use the Sum Multi|(p1, p2) function under Fitting > Edit Custom Model, however I am not able to find this model in the options list.
> 
> Checked by SMK. He is correct.
> 
> What is more, the same bug exists in v3.0.0 and v2.2.1!!!
> 
> PDB checked and it appears that no multiplicity models appear in the dropdown (eg. RPA10, onion, etc).  This is probably working as designed since I suspect the interface was not obvious for a multiplicity model?  Interestingly there is also a very strange behavior that the shape independent -> Absolute_Power_Law seems to have been renamed in this dropdown to PowerLawAbsModel ... perhaps due to the space in the name? 


* **butler** changed priority from "critical" to "major"
* **butler** changed summary from "UnifiedPowerRg model does not appear in drop-down in Custom Model > Sum|Multi" to "Multiplicity models do not appear in drop-down in Custom Model > Sum|Multi"

**Comment by butlerpd** [comment]
Trac update at `2015/12/02 02:59:47`: 
* **butler** commented:

The way this is currently handled is in !ModelManagerBase in models.py.  every model is explicitly added to a variety of lists and dictionaries.  The sum model dropdown uses the model_name_list to populate itself.  !ModelManagerBase has two mutually exclusive lists: model_name_list and multi_func_list.  This last is the list of all models which have a multiplicity parameter (selectable number of levels).  One CAN force these to also be members of model_name_list but the sum model editor  will then only produce a sum model with 1 level as there is currently no way to select that.  This would require a GUI change in src.sas.perspectives.calculator.model_editor.!TextDialog to handle.  

We will have to deal with this as part of sasmodels.


* **butler** changed priority from "major" to "critical"
* **butler** changed workpackage from "SasView Bug Fixing" to "SasModels Redesign"

**Comment by butlerpd** [comment]
Trac update at `2015/12/02 03:16:45`: 
* **butler** commented:

with regard to the naming issue it appears that the sum model editor uses the file name (!PowerLawAbsModel.py) while the fit panel's drop down uses the self.name parameter of the model.  These seem to usually be the same or similar except in the !PowerLawAbsModel which is alphebetically far away from the otherwise first Absolute Power_Law. 

The GUI should be consistent and always use one or the other.  This should be addressed as part of the new sasmodels implementation as well though it may be appropriate to divide these into two tickets


* **butler** changed description from:

> Reported by Saeed Mohan (saeed.mohan@ipleiria.pt):
> 
> I was attempting to add in my own custom model to Sasview, which is based on the !UnifiedPowerRg model. I thought the easiest thing to do would be to use the Sum Multi|(p1, p2) function under Fitting > Edit Custom Model, however I am not able to find this model in the options list.
> 
> Checked by SMK. He is correct.
> 
> What is more, the same bug exists in v3.0.0 and v2.2.1!!!
> 
> PDB checked and it appears that no multiplicity models appear in the dropdown (eg. RPA10, onion, etc).  This is probably working as designed since I suspect the interface was not obvious for a multiplicity model?  Interestingly there is also a very strange behavior that the shape independent -> Absolute_Power_Law seems to have been renamed in this dropdown to PowerLawAbsModel ... perhaps due to the space in the name? 

to:

> Reported by Saeed Mohan (saeed.mohan@ipleiria.pt):
> 
> I was attempting to add in my own custom model to Sasview, which is based on the !UnifiedPowerRg model. I thought the easiest thing to do would be to use the Sum Multi|(p1, p2) function under Fitting > Edit Custom Model, however I am not able to find this model in the options list.
> 
> Checked by SMK. He is correct.
> 
> What is more, the same bug exists in v3.0.0 and v2.2.1!!!
> 
> PDB checked and it appears that no multiplicity models appear in the dropdown (eg. RPA10, onion, etc).  This is probably working as designed since I suspect the interface was not obvious for a multiplicity model?  Interestingly there is also a very strange behavior that the shape independent -> Absolute_Power_Law seems to have been renamed in this dropdown to !PowerLawAbsModel ... perhaps due to the space in the name?

**Comment by butlerpd** [comment]
Trac update at `2016/03/20 23:13:25`: 
* **butler** commented:

naming should be consistant as of 4.0, but multiplicity models will not be able to sum/multiply for this release.  Question of whether that is something we do.


* **butler** changed milestone from "SasView 4.0.0" to "SasView Next Release +1"
* **butler** changed priority from "critical" to "major"

**Comment by butlerpd** [comment]
Trac update at `2017/10/26 12:35:20`: **butler** changed milestone from "SasView Next Release +1" to "SasView 5.1.0"

**Comment by butlerpd** [comment]
Trac update at `2019/03/02 00:16:28`: **butler** changed workpackage from "SasModels Redesign" to "SasModels Infrastructure"

**Comment by rozyczko** [comment]
Interestingly, this works in 5.0 and the product/sum model contains all relevant multiplicity combo boxes.

**Comment by smk78** [comment]
So it does! Although, if I add onion+onion I only get one of the n_shell combo boxes and the Show SLD Profile calculation fails.

**Comment by rozyczko** [comment]
Ah! Then you found a bug with the implementation. I'll fix it for 5.0.1 then.

**Comment by pkienzle** [comment]
onion+onion will not work with the current sasview wrapper.  The underlying sasmodel definition may have enough information to keep the two "n" separate (haven't tested), but the old API only allows a single selector.

Rather than hacking the sasview wrapper to allow this, it would be better to use the underlying sasmodels representation directly.  Need to do this anyway to keep track of which parameters come from which model.

**Comment by rozyczko** [comment]
This looks like more than a simple fix. Moving to 5.1 then.

**Comment by krzywon** [comment]
As noted in #2647, this was fixed in v5.0.6. Closing.
